ext/spl/spl_fixedarray.c

```diff
@@ -582,6 +587,67 @@ PHP_METHOD(SplFixedArray, __wakeup)
 	}
 }
 
+PHP_METHOD(SplFixedArray, __serialize)
+{
+	spl_fixedarray_object *intern = Z_SPLFIXEDARRAY_P(ZEND_THIS);
+	zval *current, tmp;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	array_init(return_value);
+
+	/* elements */
+	array_init_size(&tmp, intern->array.size);
+
+	for (zend_long i = 0; i < intern->array.size; i++) {
+		current = &intern->array.elements[i];
+		zend_hash_next_index_insert(Z_ARRVAL(tmp), current);
+		Z_TRY_ADDREF_P(current);
+	}
+
+	zend_hash_next_index_insert(Z_ARRVAL_P(return_value), &tmp);
+
+	/* members */
+	ZVAL_ARR(&tmp, zend_proptable_to_symtable(
+		zend_std_get_properties(&intern->std), /* always_duplicate */ 1));
+	zend_hash_next_index_insert(Z_ARRVAL_P(return_value), &tmp);
+}
+
+PHP_METHOD(SplFixedArray, __unserialize)
+{
+	spl_fixedarray_object *intern = Z_SPLFIXEDARRAY_P(ZEND_THIS);
+	HashTable *data;
+	zval *storage_zv, *members_zv, *elem;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "h", &data) == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	if (intern->array.size == 0) {
+		storage_zv = zend_hash_index_find(data, 0);
+		members_zv = zend_hash_index_find(data, 1);
+		if (!storage_zv || !members_zv ||
+			Z_TYPE_P(storage_zv) != IS_ARRAY || Z_TYPE_P(members_zv) != IS_ARRAY
+		) {
+			zend_throw_exception(spl_ce_UnexpectedValueException,
+				"Incomplete or ill-typed serialization data", 0);
+			RETURN_THROWS();
+		}
+
+		zend_long size = zend_hash_num_elements(Z_ARRVAL_P(storage_zv));
+		spl_fixedarray_init_non_empty_struct(&intern->array, size);
+
+		zend_long i = 0;
+		ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(storage_zv), elem) {
```

Comment:
`ZEND_HASH_FOREACH_KEY_VAL(ht, _h, _key, _val)` or `ZEND_HASH_FOREACH_STR_KEY_VAL(ht, _key, _val)` would help in setting this.

The values where `_val != null` are the member properties to add.

Any of the options on their own should work.
1. the array can be reallocated with `erealloc` then changing the size (check for null)
2. or manually set the remaining values (i..size-1) to ZVAL_NULL(...) before calling the helper function to resize the array normally - this should be fine since it's rare to need to resize these. 
3. Or do a second foreach loop first to compute the count of array values that have non-string keys to use as `size`

Reply:
I don't really understand what the problem is here: during serialization, all `spl_fixedarray->elements` items are exported. During deserialization, the the number of items in the serialized array (`member_zv`) is assigned to `spl_fixedarray->size`, and then we iterate through the items of `member_zv`.

What do I need to use the array key for? What did I miss?

Reply:
OK, I've just understood it. This comment relates to the backward compatible implementation of __serialize()/__unserialize().

Reply:
I think I managed to implement something similar to your option 1. I hope that I didn't miss anything..